dpnp/dpnp_iface_histograms.py

```diff
@@ -306,6 +312,61 @@ def digitize(x, bins, right=False):
     return bins.size - dpnp.searchsorted(bins[::-1], x, side=side)
 
 
+def _find_supported_dtype(dt, supported):
+    if dt in supported:
+        return dt
+
+    for t in supported:
+        if dpnp.can_cast(dt, t):
+            return t
+
+    return None
+
+
+def _result_type_for_device(dtype1, dtype2, device):
+    rt = dpnp.result_type(dtype1, dtype2)
+    return map_dtype_to_device(rt, device)
+
+
+def _align_dtypes(a_dtype, bins_dtype, ntype, device):
+    has_fp64 = device.has_aspect_fp64
+    a_bin_dtype = _result_type_for_device(a_dtype, bins_dtype, device)
+
+    supported_types = (dpnp.float32, dpnp.int64, numpy.uint64, dpnp.complex64)
```

Comment:
Wouldn't it be helpful to return a mapper function based on `ContigFactory`? Like it's done for ufunc, and then it will only need to check if we can cast the input arrays to expected matching dtype if any. And it will return the dtype of result histogram array we have to allocate.

Reply:
It more complicated than for ufunc and elementwise.
`hist_dtype` is not directly depends on `a_bin_dtype`.
And supported types matrix is not cartesian product of `a_bin_dtype` and `hist_dtype` supported types.

So types selection is quite complex and is hard to implemented in generic way.
But anyway I've added functionality to return supported types combination to python and implement some version of types selection which is ok for histogram.